### ISSUE TITLE
Gitlab CI builds not working on expo client app

### ISSUE THREAD

**Issue by zanona** [expo/expo-cli#2223]
## Description

builds published through Gitlab CI won't open on Expo client app with the error message added below.
However, when publishing it locally the builds works just fine. 

## Expected Behavior

I expect the CI build to run on Expo client app without errors 

## Observed Behavior

```
Uncaught Error: Invariant Violation: Module AppRegistry is not a registered callable module (calling runApplication),stack:
value@676:3544
<unknown>@676:841
value@676:2939
value@676:813
value@-1
```

### .gitlab-ci.yml
```yml
deploy:
  stage: deploy
  only:
    refs:
      - schedules
      - nightly
  script:
    - apk add git
    - yarn semantic-release
    - echo fs.inotify.max_user_instances=524288 | tee -a /etc/sysctl.conf && sysctl -p
    - echo fs.inotify.max_user_watches=524288 | sudo tee -a /etc/sysctl.conf && sysctl -p
    - echo fs.inotify.max_queued_events=524288 | sudo tee -a /etc/sysctl.conf && sudo sysctl
    - expo login --non-interactive -u "$EXPO_USERNAME"
    - expo publish --non-interactive --release-channel "$CI_COMMIT_REF_SLUG"
```

## Gitlab logs

```
$ expo login --non-interactive -u "$EXPO_USERNAME"
┌─────────────────────────────────────────────────────────────────────────┐
│                                                                         │
│   There is a new version of expo-cli available (3.21.3).                │
│   You are currently using expo-cli 3.19.2                               │
│   Install expo-cli globally using the package manager of your choice;   │
│   for example: `npm install -g expo-cli` to get the latest version      │
│                                                                         │
└─────────────────────────────────────────────────────────────────────────┘
Success. You are now logged in as REDACTED.
$ expo publish --non-interactive --release-channel "$CI_COMMIT_REF_SLUG"
┌─────────────────────────────────────────────────────────────────────────┐
│                                                                         │
│   There is a new version of expo-cli available (3.21.3).                │
│   You are currently using expo-cli 3.19.2                               │
│   Install expo-cli globally using the package manager of your choice;   │
│   for example: `npm install -g expo-cli` to get the latest version      │
│                                                                         │
└─────────────────────────────────────────────────────────────────────────┘
[19:30:37] Unable to find an existing Expo CLI instance for this directory; starting a new one...
[19:30:48] Starting Metro Bundler on port 19001.
[19:30:51] Publishing to channel 'nightly'...
[19:31:08] Building iOS bundle
[19:34:58] Building Android bundle
[19:34:58] Finished building JavaScript bundle in 176771ms.
[19:37:36] Analyzing assets
[19:37:38] Finished building JavaScript bundle in 159713ms.
[19:37:43] Finished building JavaScript bundle in 4571ms.
[19:37:47] Finished building JavaScript bundle in 3178ms.
[19:37:47] Uploading assets
[19:37:47] No assets changed, skipped.
[19:37:47] Processing asset bundle patterns:
[19:37:47] - /builds/app/marketplace/assets/*
[19:37:49] Building sourcemaps
[19:37:49] Finished building JavaScript bundle in 315ms.
[19:37:50] Finished building JavaScript bundle in 267ms.
[19:37:50] Uploading JavaScript bundles
...
[19:37:54] Running postPublish hook: sentry-expo/upload-sourcemaps
[19:37:54] No config found in app.json, falling back to environment variables...
[19:37:54] Created release 1.0.0-nightly.1-r.JBIsIJWkT.
[19:37:58] > Analyzing 4 sources
> Rewriting sources
> Adding source map references
> Bundled 4 files for upload
> Uploaded release files to Sentry
> File upload complete
Source Map Upload Report
  Minified Scripts
    ~/main.android.bundle (sourcemap at main.android.map)
    ~/main.ios.bundle (sourcemap at main.ios.map)
  Source Maps
    ~/main.android.map
    ~/main.ios.map
[19:37:58] Publish complete
[19:37:58] The manifest URL is: https://exp.host/@user/app?release-channel=nightly (​https://exp.host/@user/app/index.exp?release-channel=nightly&sdkVersion=37.0.0​). Learn more. (​https://expo.fyi/manifest-url​)
[19:37:58] The project page is: https://expo.io/@user/app?release-channel=nightly (​https://expo.io/@user/app?release-channel=nightly​). Learn more. (​https://expo.fyi/project-page​)
[19:37:58] › Closing Expo server
[19:37:58] › Stopping Metro bundler
```

**Comment by brentvatne** [comment]
you'll have to narrow this down further for us to help. expo-cli is completely unopinionated about where it is running. you might want to try following the **exact same process** that you are using on ci manually on a vm or locally

**Comment by zanona** [comment]
Hi, both locally and on the GitLab the build process is running inside the same docker image. 

However I could narrow it down to the node modules installation which locally I had with a simple `yarn` as on CI I had with `yarn --prefer-offline --cache-folder .yarn`, once I have aligned both to the latter, I could see that I was indeed getting the same error even when in dev mode locally. I could spot a few paths listed in the error pointing to the yarn cache folder `.yarn`.

After reverting to a simpler module installation locally and on GitLab CI with only `yarn` _(no cache folder)_, I was able to get the published apps to work.

So now the question is if there's any strict dependency on where the node modules are cached, as it's common practice on CI to point that to a different directory so it can be reused across jobs?

As an alternative I also tried setting up an env variable `YARN_CACHE_FOLDER=.yarn` which lead me into the same error.

**Comment by brentvatne** [comment]
expo-cli also isn't opinionated about where the yarn cache is, as long as yarn is able to install the node_modules correctly :)

**Comment by zanona** [comment]
Hi, @brentvatne thanks for following up.
It actually turns out yarn was behaving very oddly whenever setting a custom location for its cache and not really working. After testing it thoroughly I can only say this was indeed a dependency issue and not expo or expo-cli's. I have been able to get it working by switching to npm in the end, which in this case seemed a bit more reliable.

Closing this now, sorry for the false alarm.
